orderbook/src/orderbook.rs

```diff
@@ -516,4 +530,44 @@ mod tests {
             .unwrap();
         assert_eq!(result, &orders[1..2]);
     }
+
+    #[test]
+    fn detects_orders_with_same_buy_and_sell_token() {
+        let native_token = dummy_contract!(WETH9, [0xef; 20]);
+        assert!(has_same_buy_and_sell_token(
+            &OrderCreation {
+                sell_token: H160([0x01; 20]),
+                buy_token: H160([0x01; 20]),
+                ..Default::default()
+            },
+            &native_token,
+        ));
+        assert!(has_same_buy_and_sell_token(
+            &OrderCreation {
+                sell_token: native_token.address(),
+                buy_token: BUY_ETH_ADDRESS,
+                ..Default::default()
+            },
+            &native_token,
+        ));
+
+        assert!(!has_same_buy_and_sell_token(
+            &OrderCreation {
+                sell_token: H160([0x01; 20]),
+                buy_token: H160([0x02; 20]),
+                ..Default::default()
+            },
+            &native_token,
+        ));
+        // Sell token set to 0xeee...eee has no special meaning, so it isn't
+        // considered buying and selling the same token.
```

Comment:
With the same argument you can argue that `has_same_buy_and_sell_token(BUY_ETH_ADDRESS, BUY_ETH_ADDRESS)` should return false. Maybe we should reject any order that uses`BUY_ETH_ADDRESS` as the sell token in a different check, since it's almost certainly going to be user error.

Reply:
This should already be the case though, as there should be an error retrieving the ERC20 token balance for address `0xeeee..eeee` when verifying that the user has sufficient balance to trade it.